rfcs/import-export-api.md

```diff
@@ -0,0 +1,243 @@
+- Big reduction in Realm Proxy class size and complexity
+- Big reduction in API surface for the Realm class
+- Move towards a opt-in plugin-type architecture.
+
+**Disadvantages**
+- Harder to discover functionality. Must visit website to find it.
+- Harder to manage versions, e.g. which version of `RealmJson` is compatible
+  with Realm Java?
+- Confusing API that you need to use `Realm` to create the transaction, but
+  `realmJson` to insert data.
+
+
+### Proposal B
+
+* Generalize the current `insert/insertOrUpdate` API but allow you to register
+  "type-adapters" for different formats.
+* Add new `export` API.
+* Possible to use realm config block to configure plugins.
+
+**New Classes**
+* New `ObjectReader` interface that all importers must implemenent
+* Add new `ObjectWriter` interface for outputting objects to streams/files.
+* New API methods
+    
+    ```
+    // Will automatically try to detect importer based on type of object
+    realm.insert(Class, Object); 
+    realm.insertOrUpdate(Class, Object)
+    // Disambiguate if needed, e.g String could both be JSON or XML
+    realm.insert(Class, Object, String)
+    realm.insertOrUpdate(Class, Object, String)
+	```
+
+```java
+// build.gradle
+realm {
+	jsonSupport = true
+	xmlSupport = true
+	csvSupport = true
+	....
+}
+
+// Example
+public class MyApplication extends Application {
+
+ 	public static RealmJson realmJson;
+
+	public void onCreate() {
+		super.onCreate();
+		// Will automatically detect any of the officially supported extension
+		Realm.init();
+	}
+}
+
+// Usage
+Realm realm = Realm.getDefaultInstance();
+
+// Import
+realm.beginTransaction();
+realm.insert(Person.class, getXMLString(), "xml");
+realm.insert(Person.class, getJSONString(), "json");
+// Automatically detect which importer to use
+realm.insertOrUpdate(Person.class, getJsonStream());
+realm.commitTransaction();
+
+// Export
+RealmResults<Person> results = realm.where(Person.class).findAll();
+String data = Realm.export(results, "xml"); // Convert to String
+// Interface for serializing to whatever
+realmJson.export(results, new JsonObjectWriter("path/to/file"));
+```
+
+**Advantages**
+- API visible in main Realm API.
+- Actual implementation still fully opt-in.
+- New formats supported easily by creating `RealmXml`, `RealmCsv`, ...
+- Big reduction in Realm Proxy class size and complexity
+- Big reduction in API surface for the Realm class
+- Move towards a opt-in plugin-type architecture.
+
+**Disadvantages**
+- Methods will crash if no plugin is installed. 
+- Must visit website to find versions if not using realm config block
+- Specifying "type" as a String makes it hard to discover legal values.
+- Less flexible than a seperate class.
+
+
+## Open questions:
+
+* Can we enumerate possible input formats (XML, JSON, CSV), or does it need to
+  be flexible?
+
+* A more radical solution could also be to just remove our JSON support
+  completely and completely delegate it to 3rd party libraries?
```

Comment:
The converter has to call APIs in io.realm.internal for speed concerns. I am not quite sure we have all the checks for every internal API (but ideally we should have all checks inside the internal APIs)